File: src-backend/app/models/formulation.py
"""Pydantic models for the Optimal Formulation (最优配比) feature."""

from pydantic import BaseModel, Field


class FormulationRequest(BaseModel):
    expr_id: str
    top_k: int = Field(default=5, ge=1, le=50)
    n_samples: int = Field(default=1000, ge=100, le=10000)


class FormulationCandidate(BaseModel):
    components: dict[str, float]
    predicted_response: float
    rank: int


class FormulationResponse(BaseModel):
    expr_id: str
    candidates: list[FormulationCandidate]
    feature_names: list[str]
    attention_weights: dict[str, float]

File: src-frontend/src/types/formulation.ts
export interface FormulationRequest {
  expr_id: string;
  top_k?: number;
  n_samples?: number;
}

export interface FormulationCandidate {
  components: Record<string, number>;
  predicted_response: number;
  rank: number;
}

export interface FormulationResponse {
  expr_id: string;
  candidates: FormulationCandidate[];
  feature_names: string[];
  attention_weights: Record<string, number>;
  attention_weak?: boolean;
}
